impl/core/src/main/java/io/serverlessworkflow/impl/MonotonicUlidWorkflowInstanceIdFactory.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2020-Present The Serverless Workflow Specification Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.serverlessworkflow.impl;
+
+import de.huxhorn.sulky.ulid.ULID;
+import java.security.SecureRandom;
+
+/**
+ * A {@link WorkflowInstanceIdFactory} implementation that generates Monotonic ULIDs as workflow
+ * instance IDs.
+ */
+public class MonotonicUlidWorkflowInstanceIdFactory implements WorkflowInstanceIdFactory {
+
+  private final SecureRandom random = new SecureRandom();
+  private final ULID ulid = new ULID(random);
+  private ULID.Value previousUlid;
+
+  @Override
+  public synchronized String get() {
+    if (previousUlid == null) {
+      previousUlid = ulid.nextValue();
+    } else {
+      previousUlid = ulid.nextMonotonicValue(previousUlid);
+    }
+    return previousUlid.toString();
+  }
```

Comment:
I think this should be more lazy. 
Reason is that, with this implementaiton, if the user is replacing this factory, the secure ramdon and the ULID are still created, but never used. 
There are two possibilities to achieve that:
- creare this class, as it is, (well, secureRamdon does not need to be a member in any case) in a lazy way in the WorkflowApplication (so create it only when the user has not set another WorkflowInsanceIdFactory)
- Leave WorkflowApplication as it is and create the ULID (and the SecureRamdon) when previousUlid is null.

Reply:
I now explain why I prefer the first one (lazy initialization of the factory in WorkflowApplication)

The get method can be simplified by creating previousUlid in the constructor, so you write a not synchronized get. 

```
@Override
// note that you do not need to syncronized this
  public String get() {
       String result=  currentUlid.toString();
      currentUlid = ulid.nextMonotonicValue(currentUlid);
      return result;
  }

```
The constructor will look like

```
private final ULID ulid;
private ULID.Value currentUlid;

public MonotonicUlidWorkflowInstanceIdFactory () {
   ulid = new ULID(new SecureRandom());
   currentUlid = ulid.nextValue();
}

```